Editorial: Add "local-offset time value" for time values with local time zone offset applied

anba opened the pull request:
Commit ed3cab3b5fad06f61230baced2e0b78c5ad5eacc:

Multiple operations define their inputs as _**time values**_, but actually allow time values adjusted with a local time zone offset.

Rename _**time value**_ into _**epoch time value**_ to emphasise that it's a number of milliseconds relative to the epoch. Then add _**local time value**_ which is an _**epoch time value**_ adjusted with a time zone offset. Time zone offsets are restricted to the exclusive range ±86'400'000 milliseconds.

Finally update all references to _**time value**_ to use either _**epoch time value**_ or _**local time value**_.

---

Two drive-by fixes:
- 13be0f41410f71f3898b8e8c02f35b349ca1e8cb: Ensure too large inputs are properly handled.
- e547c587663d542c4007d688d769cdd4b6b7d16b: Restrict time zone offsets in other operations (fixes #3101).

anba committed:
Editorial: Add "local-offset time value" for time values with local time zone offset applied

Multiple operations define their inputs as "time values", but actually
allow time values adjusted with a local time zone offset.

Add "local-offset time value" which is a "time value" adjusted with a time
zone offset. Time zone offsets are restricted to the exclusive range
±86'400'000 milliseconds.

anba committed:
Editorial: Handle too large inputs early in UTC

Return `NaN` if `t` exceeds the "local time value" range, because
operations called within `UTC` expect their arguments to be "local time
values" and not arbitrarily large integers.

anba committed:
Editorial: Restrict GetNamedTimeZoneOffsetNanoseconds return values less than 24 hours

Fixes #3101

anba commented:
cc @ptomato

gibson042 commented on spec.html:
```diff
@@ -32580,18 +32580,19 @@ <h1>Date Objects</h1>
+        <p>Epoch time values do not account for UTC leap seconds—there are no epoch time values representing instants within positive leap seconds, and there are epoch time values representing instants removed from the UTC timeline by negative leap seconds. However, the definition of epoch time values nonetheless yields piecewise alignment with UTC, with discontinuities only at leap second bounda ...
+        <p>A Number can exactly represent all integers from -9,007,199,254,740,992 to 9,007,199,254,740,992 (<emu-xref href="#sec-number.min_safe_integer"></emu-xref> and <emu-xref href="#sec-number.max_safe_integer"></emu-xref>). An epoch time value supports a slightly smaller range of -8,640,000,000,000,000 to 8,640,000,000,000,000 milliseconds. This yields a supported epoch time value range of ...
+        <p>The exact moment of midnight at the beginning of 1 January 1970 UTC is represented by the epoch time value *+0*<sub>𝔽</sub>.</p>
+        <p>An ECMAScript <dfn variants="local time values">local time value</dfn> is an epoch time value adjusted by a time zone offset. Time zone offsets are integers in the exclusive range -86,400,000 to 86,400,000.</p>
```
Two comments here:
1. We have been missing a good explanation of "local" for a while, and this is a good place to include it.
2. "[interval](https://tc39.es/ecma262/multipage/notational-conventions.html#interval)" is probably better than "exclusive range".

Feel free to tweak:
```suggestion
        <p>An ECMAScript <dfn variants="local time values">local time value</dfn> is an epoch time value adjusted by a time zone offset that is an integer number of milliseconds in the interval from -86,400,000 (exclusive) to 86,400,000 (exclusive). A local time value represents a local date and time of day, specifically the UTC date and time of day corresponding an identical epoch time value, ignoring range limits. For example, a local time value of *60000*<sub>𝔽</sub> represents time of day 00:01 on 1 January 1970, and could equivalently come from any of the following:</p>
        <ul>
          <li>local offset +00:00 = *+0*<sub>𝔽</sub> at epoch time 1970-01-01T00:01Z = *60000*<sub>𝔽</sub></li>
          <li>local offset +01:00 = *3600000*<sub>𝔽</sub> at epoch time 1969-12-31T23:01Z = 𝔽(-msPerHour + 60000) = *-3540000*<sub>𝔽</sub></li>
          <li>local offset -05:00 = 𝔽(-5 × msPerHour) = *-18000000*<sub>𝔽</sub> at epoch time 1970-01-01T05:01Z = 𝔽(5 × msPerHour + 60000) = *18060000*<sub>𝔽</sub></li>
        </ul>
```

anba replied:
Added as 81aa87e2c8ea1419ba844a4383245836c5ad4ce1, can later be merged with 11579983a9f6c51d6c7b3adc22557dc4ff29381c.

gibson042 reviewed (approved):
I love it!

anba commented on spec.html:
```diff
@@ -52042,10 +52044,10 @@ <h1>Built-in Objects and Methods</h1>
   <p><emu-xref href="#sec-candeclareglobalvar"></emu-xref>-<emu-xref href="#sec-createglobalfunctionbinding"></emu-xref> Edition 5 and 5.1 used a property existence test to determine whether a global object property corresponding to a new global declaration already existed. ECMAScript 2015 uses an own property existence test. This corresponds to what has been most commonly implemented by web brow ...
   <p><emu-xref href="#sec-array-exotic-objects-defineownproperty-p-desc"></emu-xref>: The 5<sup>th</sup> Edition moved the capture of the current array length prior to the integer conversion of the array index or new length value. However, the captured length value could become invalid if the conversion process has the side-effect of changing the array length. ECMAScript 2015 specifies that the c ...
-  <p><emu-xref href="#sec-timeclip"></emu-xref>: Previous editions permitted the TimeClip abstract operation to return either *+0*<sub>𝔽</sub> or *-0*<sub>𝔽</sub> as the representation of a 0 time value. ECMAScript 2015 specifies that *+0*<sub>𝔽</sub> always returned. This means that for ECMAScript 2015 the time value of a Date is never observably *-0*<sub>𝔽</sub> and methods that return time val ...
+  <p><emu-xref href="#sec-timeclip"></emu-xref>: Previous editions permitted the TimeClip abstract operation to return either *+0*<sub>𝔽</sub> or *-0*<sub>𝔽</sub> as the representation of a 0 epoch time value. ECMAScript 2015 specifies that *+0*<sub>𝔽</sub> always returned. This means that for ECMAScript 2015 the epoch time value of a Date is never observably *-0*<sub>𝔽</sub> and methods that ret ...
```
TimeClip can be called with any Number input. Only the output is guaranteed to be an epoch time value.

MakeDate has an unbounded output, so we can't use epoch or local time value.

gibson042 replied:
I'd still like to see at least an editorial note explaining the above, then. Right now those operations are somewhat orphaned.

anba replied:
I've added e4aec515b51dfc847f78a8ddae868a060c1c1875 and 6d7452a49b5f11558e5a256453a3203135703d93 to give MakeDay, MakeTime, MakeDate, and TimeClip more narrow input/output types.

bakkot commented:
This is a great change. Editors would like to submit for consideration "time value" and "local-offset time value" as the names; @syg found "epoch time value" and "local time value" confusing because "local time value" is also based on the time since the epoch.

anba committed:
Editorial: Narrow input and output types for MakeDay, MakeTime, and MakeDate

anba committed:
Editorial: Narrow output type for TimeClip

anba committed:
Fixup: Review suggestion from gibson042

anba commented:
> Editors would like to submit for consideration "time value" and "local-offset time value" as the names;

Sure. I've updated b5b1f0aba53627c2be6e6b4899fc12f4e4d7e911 to use the suggested names. 

Furthermore I've added e4aec515b51dfc847f78a8ddae868a060c1c1875 and 6d7452a49b5f11558e5a256453a3203135703d93 to narrow some input resp. output types. And added 03ddef504268287d2b8e1316e6ba5766e6634457 per the above review suggestion.

michaelficarra commented on spec.html:
```diff
@@ -32589,6 +32589,12 @@ <h1>Time Values and Time Range</h1>
         <p>Time values do not account for UTC leap seconds—there are no time values representing instants within positive leap seconds, and there are time values representing instants removed from the UTC timeline by negative leap seconds. However, the definition of time values nonetheless yields piecewise alignment with UTC, with discontinuities only at leap second boundaries and zero difference ...
         <p>A Number can exactly represent all integers from -9,007,199,254,740,992 to 9,007,199,254,740,992 (<emu-xref href="#sec-number.min_safe_integer"></emu-xref> and <emu-xref href="#sec-number.max_safe_integer"></emu-xref>). A time value supports a slightly smaller range of -8,640,000,000,000,000 to 8,640,000,000,000,000 milliseconds. This yields a supported time value range of exactly -100 ...
         <p>The exact moment of midnight at the beginning of 1 January 1970 UTC is represented by the time value *+0*<sub>𝔽</sub>.</p>
+        <p>An ECMAScript <dfn variants="local-offset time values">local-offset time value</dfn> is a time value adjusted by a time zone offset that is an integer number of milliseconds in the interval from -86,400,000 (exclusive) to 86,400,000 (exclusive). A local-offset time value represents a local date and time of day, specifically the UTC date and time of day corresponding an identical time v ...
```
We talked in editor call about how technically a local-offset time value isn't a time value and that might be confusing to readers. But I don't really mind explaining it like this. @tc39/ecma262-editors Thoughts?

michaelficarra commented on spec.html:
```diff
@@ -33131,7 +33138,7 @@ <h1>
         </emu-alg>
         <p>
-          Input _t_ is nominally a time value but may be any Number value.
+          Input _t_ is nominally a local-offset time value but may be any Number value.
           The algorithm must not limit _t_ to the time value range, so that inputs corresponding with a boundary of the time value range can be supported regardless of local UTC offset.
           For example, the maximum time value is 8.64 × 10<sup>15</sup>, corresponding with *"+275760-09-13T00:00:00Z"*.
           In an environment where the local time zone offset is ahead of UTC by 1 hour at that instant, it is represented by the larger input of 8.64 × 10<sup>15</sup> + 3.6 × 10<sup>6</sup>, corresponding with *"+275760-09-13T01:00:00+01:00"*.
```
I suggest we redefine local-offset time value to account for this additional range, then type this AO properly.

michaelficarra replied:
Alternatively, I don't think it'd be a problem to just expand the range of time value either, right?

anba replied:
`UTC` is called with unbounded integral Number values, so we can't restrict the input to time values.

michaelficarra replied:
Okay. I still think we should expand the range supported by local-offset time values.

Also, it's unrelated to this PR, but: `t` isn't *nominally* a time value, it's *nominally* a Number. It's "ostensibly" or "figuratively" a time value. I would update that term as well.

anba replied:
I don't understand to which range local-offset time values should be extended? Any integral Number, including NaN (and possibly Infinity, too)? Because that's the range which can be produced by MakeDate. (And the output of MakeDate is passed as the input to UTC.)

michaelficarra replied:
ISO 8601 timezone offsets permit ±24 hours (less 1 nanosecond). Time values are integers within the range -8,640,000,000,000,000 to 8,640,000,000,000,000 (100 million days on either side of the epoch). I'm suggesting that local time values have an extended range of ±8,640,000,086,400,000 (100,000,001 days on either side of the epoch).

anba replied:
±8,640,000,086,400,000 suggests it's valid for a time zone offset to be ±24 hours, but just as you said, the maximum valid time zone offset is ±23:59:59.999.999.999 hours. Ignoring micro- and nanoseconds, we get ±23:59:59.999, which is ±86399999 = ±(86400000 - 1) milliseconds, so exactly the range already used within this PR.

michaelficarra replied:
Can you point to where a local-offset time value is defined in that way in this PR? All I see is that it's restricted to the same range as un-offset time values (±8,640,000,000,000,000).

anba replied:
Ah, I think I understand our misunderstanding.

The first commit has:
> An ECMAScript <dfn variants="local-offset time values">local-offset time value</dfn> is a time value adjusted by a time zone offset that is an integer number of milliseconds in the interval from -86,400,000 (exclusive) to 86,400,000 (exclusive).

I guess "time value adjusted by a time zone offset" can be read in two different ways:
1. Take a time value ±(8.64 * 10^15) and adjust it by ±86399999 to get ±8,640,000,086,399,999 as the final range.
2. "local-offset time value" is still a "time value", so restricted to ±(8.64 * 10^15), but additionally has some time zone offset applied.

michaelficarra replied:
Yes, I read it the second way. I would prefer to explicitly state the final computed range in its definition.

michaelficarra commented on spec.html:
```diff
@@ -33069,7 +33075,7 @@ <h1>SystemTimeZoneIdentifier ( ): a String</h1>
           LocalTime (
             _t_: a finite time value,
-          ): an integral Number
+          ): a local-offset time value
```
If, for some reason, we *don't* make all local-offset time values finite, we should add the finite qualifier here.

michaelficarra commented on spec.html:
```diff
@@ -32966,7 +32972,7 @@ <h1>
             _timeZoneIdentifier_: a String,
             _epochNanoseconds_: a BigInt,
-          ): an integer
+          ): an integer in the exclusive interval from -8.64 × 10<sup>13</sup> to 8.64 × 10<sup>13</sup>
```
I would prefer defining this type up in Time Values and Time Range alongside "local-offset time value" as something like "timezone offset".

ptomato replied:
We can reuse that definition in Temporal.

michaelficarra commented on spec.html:
```diff
@@ -32966,7 +32972,7 @@ <h1>
             _timeZoneIdentifier_: a String,
             _epochNanoseconds_: a BigInt,
-          ): an integer
+          ): an integer in the exclusive interval from -8.64 × 10<sup>13</sup> to 8.64 × 10<sup>13</sup>
```
We actually don't define the term "exclusive interval". We should add a `dfn` for it.

ptomato commented on spec.html:
```diff
@@ -32695,7 +32701,7 @@ <h1>
         <h1>
           YearFromTime (
-            _t_: a finite time value,
+            _t_: a finite local-offset time value,
```
Note that YearFromTime can still be called from MakeDay with a Number that is outside of the local-offset time value range, for example in `Date.UTC(-271821, 3, 20)`. Explanation in https://github.com/tc39/ecma262/issues/1087#issuecomment-1120110847. Same for MonthFromTime and DateFromTime. (This is pre-existing, so doesn't need to block the PR.)

ptomato commented:
Glad to see this!
